src/platform/mt8186/lib/clk.c

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2022 Mediatek. All rights reserved.
+//
+// Author: Allen-KH Cheng <allen-kh.cheng@mediatek.com>
+//         Tinghan Shen <tinghan.shen@mediatek.com>
+
+#include <platform/drivers/mt_reg_base.h>
+#include <sof/common.h>
+#include <sof/lib/clk.h>
+#include <sof/lib/cpu.h>
+#include <sof/lib/io.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/notifier.h>
+#include <sof/lib/uuid.h>
+#include <sof/sof.h>
+#include <sof/spinlock.h>
+#include <sof/trace/trace.h>
+
+DECLARE_SOF_UUID("clkdrv", clkdrv_uuid, 0x23b12fd5, 0xc2a9, 0x41a8,
+		 0xa2, 0xb3, 0x23, 0x1a, 0xb7, 0xdc, 0xdc, 0x70);
+
+DECLARE_TR_CTX(clkdrv_tr, SOF_UUID(clkdrv_uuid), LOG_LEVEL_INFO);
+
+/* default voltage is 0.8V */
+const struct freq_table platform_cpu_freq[] = {
+	{  26000000, 26000},
+	{ 100000000, 26000},
+	{ 200000000, 26000},
+	{ 400000000, 26000},
+	{ 800000000, 26000},
+};
+
+STATIC_ASSERT(ARRAY_SIZE(platform_cpu_freq) == NUM_CPU_FREQ,
+	      invalid_number_of_cpu_frequencies);
+
+static SHARED_DATA struct clock_info platform_clocks_info[NUM_CLOCKS];
+
+static void set_mux_sel(enum mux_id_t mux_id, uint32_t value)
+{
+	switch (mux_id) {
+	case MUX_CLK_ADSP_SEL:
+		io_reg_write(MTK_CLK_CFG_11_CLR,  (MTK_CLK_ADSP_MASK << MTK_CLK_ADSP_OFFSET));
+		io_reg_write(MTK_CLK_CFG_11_SET,  (value << MTK_CLK_ADSP_OFFSET));
+		io_reg_write(MTK_CLK_CFG_UPDATE,  MTK_CLK_CFG_ADSP_UPDATE);
+
+		tr_dbg(&clkdrv_tr, "adsp_clk_mux=%x, CLK_CFG_11=0x%08x\n",
+		       value, io_reg_read(MTK_CLK_CFG_11));
+		break;
+	case MUX_CLK_ADSP_BUS_SRC_SEL:
+		io_reg_write(MTK_ADSP_BUS_SRC, value);
+		io_reg_write(MTK_ADSP_CLK_BUS_UPDATE, MTK_ADSP_CLK_BUS_UPDATE_BIT);
+
+		tr_dbg(&clkdrv_tr, "adsp_bus_mux=%x, MTK_ADSP_BUS_SRC=0x%08x\n",
+		       value, io_reg_read(MTK_ADSP_BUS_SRC));
+		break;
```

Comment:
there's no shared code between the two muxes and the caller always uses a constant for the `mux_id`, maybe just make two separate functions?